### DISCUSSION TITLE
No syntax highlight for C++ with tree-sitter

### DISCUSSION THREAD

**Discussion by emacsbliss** [abougouffa/minemacs#135] [Q&A] [answered]
When I use C++-ts-mode and installed language grammar, open C++ file there is no syntax highlighting for most parts, only for `#include`. 

Does that mean I need to install some extra package to get that? Looks like `c++-mode` has syntax highlight but that's not tree-sitter based.

**Comment by abougouffa** [comment]
Which version of Emacs you are using? Try to evaluate `system-configuration-features` to see if Emacs has been built with tree-sitter support.

**Reply by emacsbliss** [reply]
emacs version: GNU Emacs 29.0.92 (build 2, aarch64-apple-darwin22.5.0, NS appkit-2299.60 Version 13.4.1 (Build 22F82)) of 2023-07-19

`system-configuration-features` has value:
```
"ACL GLIB GNUTLS IMAGEMAGICK JPEG JSON LCMS2 LIBXML2 MODULES NATIVE_COMP NOTIFY KQUEUE NS PDUMPER PNG RSVG SQLITE3 THREADS TOOLKIT_SCROLL_BARS TREE_SITTER WEBP XIM XWIDGETS ZLIB"
```

**Reply by abougouffa** [reply]
Can you try `M-x minemacs-run-build-functions` or `M-x treesit-auto-install-all` ?

**Comment by donneyluck** [comment]
i have same question ~
has TREE_SITTER too

**Reply by abougouffa** [reply]
I've faced the same issue (Emacs not loading the treesitter*.so files correctly. I've fixed it by recompiling Emacs (or maybe, reinstalling Emacs if you have installed it from your distro's package manager).

Then, by running:

```shell
cd ~/.emacs.d  # Where MinEmacs is cloned
make clean_all
emacs
```

**Reply by donneyluck** [reply]
What is your emacs version? I used the latest version of emacs on manjaro and reinstalled and compiled it. It still cannot be displayed normally.

**Reply by abougouffa** [reply]
I'm using a locally compiled version of emacs-29 branch. Emacs v30 feels a bit unstable for me, it should work with the config, but I didn't test it exhaustively.

**Reply by donneyluck** [reply]
Ok I will test the local compiled version later. Can you tell me your compilation options and parameters?

**Reply by donneyluck** [reply]
I recompiled locally and installed 29.3 and the problem was solved  thks

**Comment by donneyluck** [comment]
After I reinstalled the new version of minemacs, I can no longer use c++-ts-mode normally. I have a headache.

**Reply by abougouffa** [reply]
Weird, C/C++ are the modes that I use the most, and I didn't notice anything wrong with it lately.

Can you try with the last commit? I suspect an external package is causing this.

**Reply by donneyluck** [reply]
> Weird, C/C++ are the modes that I use the most, and I didn't notice anything wrong with it lately.
> 
> Can you try with the last commit? I suspect an external package is causing this.

I recompiled emacs 29.3 and completely reinstalled minemacs, but c++-ts-mode still doesn't display normally. It's a real headache.

emacs 29.3 ./configure --with-native-compilation=aot --with-tree-sitter
manjaro i3wm lastest
minemacs lastest

**Reply by abougouffa** [reply]
Do you have the tree-sitter grammars installed? try `treesit-auto-install-all` to be sure.

I don't see why you have this behavior since I'm using MinEmacs mainly for C/C++ development and didn't notice something like this.

You can also try the `next` branch of MinEmacs!

**Reply by donneyluck** [reply]
> Do you have the tree-sitter grammars installed? try `treesit-auto-install-all` to be sure.
> 
> I don't see why you have this behavior since I'm using MinEmacs mainly for C/C++ development and didn't notice something like this.
> 
> You can also try the `next` branch of MinEmacs!

libtree-sitter-cpp.so exists

I re-clone the next branch and it still cannot display colors correctly. I will test other versions of emacs later.

**Reply by abougouffa** [reply]
@donneyluck I've got the same behavior on my second machine. I can see that only some keywords are higlighted (only `true`, `false` and `nullptr` so far).

![Screenshot_20240604_231442](https://github.com/abougouffa/minemacs/assets/3716399/bfeee7cc-b318-4204-9aa0-ebd175a0bd28)

I don't know the cause of this issue, I will try to reproduce it in a minimal example to see if it comes from MinEmacs customization or from Emacs or Tree-sitter.

**Reply by abougouffa** [reply]
I've tested on a vanilla Emacs with `emacs -Q`. The issue is present also, so it is not a problem from MinEmacs!

Evaluate:
```elisp
(setq treesit-language-source-alist '((cpp "https://github.com/tree-sitter/tree-sitter-cpp" nil nil nil nil)))
(treesit-install-language-grammar 'cpp)
```

Then `C-x C-f` and open a C++ file.

![image](https://github.com/abougouffa/minemacs/assets/3716399/57d833c3-c0ab-4e36-a7e7-4392702eef85)

**Comment by abougouffa** [comment, marked as answer]
@emacsbliss @donneyluck 

I managed to find the cause of the issue. It seems that the Tree-Sitter C++ grammar causes the issue after v0.22.0. I've pushed a fix https://github.com/abougouffa/minemacs/commit/b10bca5bb180fa3b31b07f9d18e5dab5e467b754 to force `treesit-auto` to use the v0.22.0 of the grammar, and this fixes the problem for me.

You need to update MinEmacs to the last commit and remove the `~/.emacs.d/local/tree-sitter/libtree-sitter-cpp.so` file, restart Emacs, then run `treesit-auto-install-all`.

Can you please confirm that the issue is fixed for you?

### Edit
I've opened a PR in `treesit-auto` to fix the grammar revision to v0.22.0 until the issue get fixed https://github.com/renzmann/treesit-auto/pull/106.

**Reply by donneyluck** [reply]
Thanks again for your help    
you are really awesome
